File: algorithms/3_hard/N/01301_number_of_paths_with_max_score/dp_rolling_array.cpp
typedef ptrdiff_t isize;

constexpr int mod = 1'000'000'007;

class Solution {
public:
    static vector<int> pathsWithMaxScore(const vector<string>& board) {
        const isize n = std::ssize(board);
        assert(n >= 2 && n == std::ssize(board[0]));

        auto buff = vector<int>(4 * (n + 1), 0);

        auto curr_sum = span<int>(&buff[0], n + 1);
        auto next_sum = span<int>(&buff[n + 1], n + 1);
        auto curr_freq = span<int>(&buff[2 * (n + 1)], n + 1);
        auto next_freq = span<int>(&buff[3 * (n + 1)], n + 1);

        for (isize i = n - 1; i >= 0; i -= 1) {
            for (isize j = n - 1; j >= 0; j -= 1) {
                if (i == n - 1 && j == n - 1) {
                    curr_freq[j] = 1;
                } else {
                    const char curr_char = board[i][j];
                    if (curr_char != 'X') {
                        curr_sum[j] = next_sum[j + 1];
                        curr_freq[j] = next_freq[j + 1];

                        if (curr_sum[j] == curr_sum[j + 1]) {
                            curr_freq[j] = (curr_freq[j] + curr_freq[j + 1]) % mod;
                        } else if (curr_sum[j] < curr_sum[j + 1]) {
                            curr_sum[j] = curr_sum[j + 1];
                            curr_freq[j] = curr_freq[j + 1];
                        }

                        if (curr_sum[j] == next_sum[j]) {
                            curr_freq[j] = (curr_freq[j] + next_freq[j]) % mod;
                        } else if (curr_sum[j] < next_sum[j]) {
                            curr_sum[j] = next_sum[j];
                            curr_freq[j] = next_freq[j];
                        }

                        if (curr_char >= '0' && curr_char <= '9') {
                            curr_sum[j] += curr_char - '0';
                        }
                    }
                }
            }

            std::swap(curr_sum, next_sum);
            std::swap(curr_freq, next_freq);
        }

        if (next_freq[0] == 0) {
            return vector<int>{0, 0};
        }

        return vector<int>{next_sum[0], next_freq[0]};
    }
};

File: algorithms/3_hard/N/01301_number_of_paths_with_max_score/dp_rolling_array.rs
impl Solution {
    const MOD_TO: i32 = 1_000_000_007;
    pub fn paths_with_max_score(board: Vec<String>) -> Vec<i32> {
        let n = board.len();
        assert!(n >= 2 && board[0].len() == n);
        let mut buff = vec![0_i32; 4 * (n + 1)];
        let (mut sum, mut freq) = buff.split_at_mut(2 * (n + 1));
        let (mut curr_sum, mut next_sum) = sum.split_at_mut(n + 1);
        let (mut curr_freq, mut next_freq) = freq.split_at_mut(n + 1);

        for i in (0..n).rev() {
            for j in (0..n).rev() {
                if i == n - 1 && j == n - 1 {
                    curr_freq[j] = 1;
                } else {
                    let curr_byte = board[i].as_bytes()[j];
                    if curr_byte != b'X' {
                        (curr_sum[j], curr_freq[j]) = (next_sum[j + 1], next_freq[j + 1]);

                        if curr_sum[j] == curr_sum[j + 1] {
                            curr_freq[j] = (curr_freq[j] + curr_freq[j + 1]) % Self::MOD_TO;
                        } else if curr_sum[j] < curr_sum[j + 1] {
                            (curr_sum[j], curr_freq[j]) = (curr_sum[j + 1], curr_freq[j + 1]);
                        }

                        if curr_sum[j] == next_sum[j] {
                            curr_freq[j] = (curr_freq[j] + next_freq[j]) % Self::MOD_TO;
                        } else if curr_sum[j] < next_sum[j] {
                            (curr_sum[j], curr_freq[j]) = (next_sum[j], next_freq[j]);
                        }

                        if curr_byte >= b'0' && curr_byte <= b'9' {
                            curr_sum[j] += (curr_byte - b'0') as i32;
                        }
                    }
                }
            }

            (curr_sum, next_sum) = (next_sum, curr_sum);
            (curr_freq, next_freq) = (next_freq, curr_freq);
        }

        if next_freq[0] == 0 {
            return vec![0, 0];
        }

        vec![next_sum[0], next_freq[0]]
    }
}
